### PR TITLE
Translate strings used in non components (Lightbox and message action sheet)

### PR DESCRIPTION
**Pull request by kunall17**
Using the react-intl inject API a '_' called function (just like the web server) is being passed on to the context and used in various locations.

### CODE REVIEW

File: src/lightbox/LightboxActionSheet.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /* @flow */
-import type { Auth } from '../types';
+import type { Auth, TranslateStringType } from '../types';
```
Comment (borisyankov) [inline src/lightbox/LightboxActionSheet.js]: We are not naming types `*Type`. Remove that part.
Also, try to think of a more descriptive name. Maybe `StringTranslator` or even `TextTranslator` (as 'text' is used in 'getText')

File: src/boot/TranslationAPIProvider.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/* @flow */
+import { PureComponent } from 'react';
+import { injectIntl, intlShape } from 'react-intl';
+
+import type { ChildrenArray } from '../types';
+
+type Props = {
+  children?: ChildrenArray<*>,
+  intl: intlShape.isRequired,
```
Comment (borisyankov) [inline src/boot/TranslationAPIProvider.js]: This does not look like a correct type.

File: src/boot/TranslationAPIProvider.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+class TranslationAPIProvider extends PureComponent<Props> {
+  static childContextTypes = {
+    _: string => {},
```
Comment (borisyankov) [inline src/boot/TranslationAPIProvider.js]: `string => {},` is not correct here - it defines a function with an argument **named** `string`. This should not reflect the function signature, it is a placeholder that we use in place of PropTypes. So just `() => {}` is enough.

Do not forget to update the `types.js` => `type Context` with the `_` value.

File: src/lightbox/LightboxActionSheet.js (hunk 1 above)
Comment (kunall17) [reply src/lightbox/LightboxActionSheet.js]: Done

File: src/boot/TranslationAPIProvider.js (hunk 3 above)
Comment (kunall17) [reply src/boot/TranslationAPIProvider.js]: Yeah, that was a typo, and in this I'm returning a string so I changed to `() => '' `

**Comment by kunall17** [comment]
Thanks for review @borisyankov, made the changes!

File: src/boot/TranslationAPIProvider.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import type { ChildrenArray } from '../types';
+
+type Props = {
+  children: ChildrenArray<*>,
+  intl: intlShape.isRequired,
+};
+
+class TranslationAPIProvider extends PureComponent<Props> {
+  static childContextTypes = {
+    _: () => '',
```
Comment (gnprice) [inline src/boot/TranslationAPIProvider.js]: I've never really quite understood what `contextTypes` is supposed to do. @borisyankov , can you explain why it's there and why the types all seem to be `() => {}`?

I did look at https://reactjs.org/docs/legacy-context.html which seems to be the relevant documentation, but the types there are all from `PropTypes`.

File: src/i18n/translations/messages_en.json [hunk 5]
```diff
@@ -137,5 +137,7 @@
   "Topics": "Topics",
   "Welcome to Zulip!\n\nYou'll need to first create an account from your computer. If you're not sure where to start, go to zulipchat.com from your web browser.\n\nHope to see you back here soon!\n": "Welcome to Zulip!\n\nYou'll need to first create an account from your computer. If you're not sure where to start, go to zulipchat.com from your web browser.\n\nHope to see you back here soon!\n",
   "Add subscribers": "Add subscribers",
-  "Only organization admins are allowed to post to this stream.": "Only organization admins are allowed to post to this stream."
+  "Only organization admins are allowed to post to this stream.": "Only organization admins are allowed to post to this stream.",
```
Comment (gnprice) [inline src/i18n/translations/messages_en.json]: Small thing, but: let's always add to this list somewhere in the middle -- really anywhere other than the very last line. That avoids unnecessary churn in the diff due to the lack of a trailing comma.

File: src/boot/TranslationAPIProvider.js (hunk 3 above)
Comment (gnprice) [reply src/boot/TranslationAPIProvider.js]: Let's keep this `() => {}` like everywhere else -- these lines are still puzzling to me :smile: , but there's no point in making it different in a way that we don't have a good story for why it'd help anything.

**Comment by kunall17** [comment]
@gnprice Updated!

**Comment by gnprice** [comment]
Hi @kunall17 -- thanks for the update!

I realized that the API I had in mind for handling translations would require some close reading of React docs and some experimentation... so I went and spent today doing that. I just pushed a series  c22dfee9b^..9eaa05c27 adding code to provide `_` and a new API for getting access to it, followed by 
a long second series c94586ca9^..d688069be making a number of refactors to the `messageActionSheet` code.

Would you try rebasing this branch on top of master, and using the new API for the lightbox code? I think that would be a good next step.

**Comment by kunall17** [comment]
@gnprice Updated the branch :)

**Comment by chrisbobbe** [comment]
This has gathered quite a few conflicts since 2018, so I'm closing it. Thanks for your efforts, @kunall17!

**Pull request closed without merging**